File: quick_sort/python/quick_sort.py
#!/usr/bin/env python3

from typing import List, Any


def _partition(arr: List[Any], lo: int, hi: int) -> int:
  # We'll take the first element in the array as our pivot.
  pivot = arr[lo]
  i, j = lo + 1, hi

  while True:
    while arr[i] < pivot:
      if i == hi: break
      i += 1
    while arr[j] > pivot:
      j -= 1

    if j <= i: break
    arr[i], arr[j] = arr[j], arr[i]

  arr[lo], arr[j] = arr[j], arr[lo]
  return j

def _quicksort(arr: List[Any], lo: int, hi: int) -> None:
  if lo >= hi:
    return
  j = _partition(arr, lo, hi)
  _quicksort(arr, lo, j - 1)
  _quicksort(arr, j + 1, hi)

def quicksort(arr: List[Any]) -> None:
  return _quicksort(arr, 0, len(arr) - 1)

File: quick_sort/c/quick_sort.c
#include <stdio.h>
#define swap(a, b) int tmp = (a); (a) = (b); (b) = tmp;

static int partition(int *arr, int lo, int hi)
{
  /* We'll take the first element in the array as our pivot. */
  int pivot = arr[lo];

  int i = lo, j = hi + 1;
  while (true) {
    while (arr[++i] < pivot) if (i == hi) break;
    while (arr[--j] > pivot); /* if (j == lo) break; */

    if (j <= i) break;
    swap(arr[i], arr[j]);
  }

  swap(arr[lo], arr[j]);
  return j;
}

static void quicksort_(int *arr, int lo, int hi)
{
  if (hi <= lo) return;
  int j = partition(arr, lo, hi);
  quicksort_(arr, lo, j - 1);
  quicksort_(arr, j + 1, hi);
}

void quicksort(int *arr, int size)
{
  return quicksort_(arr, 0, size - 1);
}
